standard/tournament_structure.lua

```diff
@@ -0,0 +1,413 @@
+--- Example of template arguments:
+--- |tournament1=PiG Sty Festival
+--- |tournament2=StayAtHome Story Cup/4#Group Stage 2
+--- |tournament3=User:(16thSq) Kuro/Master Swan Open/64
+--- |matchGroupId1=Z1lDMZPiGA
+--- |matchGroupId2=Liquipedia_wnbxUh4Vm1
+---@param args table
+---@return table?
+function TournamentStructure.readMatchGroupsSpec(args)
+	local matchGroupIds = {args.id}
+	table.insert(matchGroupIds, args.matchGroupId)
+	for _, id in Table.iter.pairsByPrefix(args, 'matchGroupId') do
+		table.insert(matchGroupIds, id)
+	end
+
+	local pageNames = {}
+	table.insert(pageNames, args.tournament)
+	for _, pageName in Table.iter.pairsByPrefix(args, 'tournament') do
+		table.insert(pageNames, pageName)
+	end
+
+	local function resolve(rawPageName)
+		local namespaceName, basePageName, stageName = TournamentStructure._splitPageName(rawPageName)
+
+		-- args.ns is deprecated
+		if not namespaceName and args.ns then
+			namespaceName = Namespace.nameFromId(args.ns)
+		end
+
+		local pageName = basePageName
+			and table.concat(Array.extend(namespaceName, basePageName), ':')
+			or FULL_PAGENAME
+		return table.concat({
+			TournamentStructure._resolveRedirect(pageName),
+			stageName
+		}, '#')
+	end
+
+	if #matchGroupIds ~= 0 or #pageNames ~= 0 then
+		return {
+			matchGroupIds = matchGroupIds,
+			pageNames = Array.map(pageNames, resolve),
+		}
+	else
+		return nil
+	end
+end
+
+function TournamentStructure.currentPageSpec()
+	return {
+		matchGroupIds = {},
+		pageNames = {FULL_PAGENAME},
+	}
+end
+
+TournamentStructure._resolveRedirect = FnUtil.memoize(function(pageName)
+	return mw.ext.TeamLiquidIntegration.resolve_redirect(pageName)
+end)
+
+--- Sorts given group tables and brackets for a given spec into stages
+---
+--- Limitations:
+--- - Stages cannot span multiple tournament pages
+--- - The stageName field (as specified by {{Stage|...}})
+--- - A stage is contigious within a page
+--- - A stage cannot consist of both matchlists and brackets
+---
+--- Stages are ordered by the match2bracketindex page variable. For stages originating from different pages,
+--- the tournamentX arg determines the ordering of pages, hence stage order.
+---@param groupTables table
+---@param brackets table
+---@param spec {matchGroupIds: table, pageNames: table}
+---@return table
+function TournamentStructure.groupByStage(groupTables, brackets, spec)
+	local function getStageKey(recordGroup)
+		return {
+			recordGroup[1].stageIndex or -1,
+			recordGroup[1].pagename,
+			TournamentStructure.getStageName(recordGroup) or 'default',
+			TournamentStructure.isGroupTable(recordGroup) and 'groupTable' or 'bracket'
+		}
+	end
+
+	local stageIndexes = Table.map(spec.pageNames, function(stageIndex, pageName) return pageName, stageIndex end)
+	local currentPage = FULL_PAGENAME
+	stageIndexes[currentPage] = stageIndexes[currentPage] or 1000
+
+	local getSortKey = FnUtil.memoize(function(recordGroup)
+		local basePageName = recordGroup[1].pagename:gsub('_', ' ')
```

Comment:
because the pagenames get redirect resolved in `TournamentStructure.readMatchGroupsSpec`
> `local redirectedPage = mw.title.new(TournamentStructure._resolveRedirect(pageName))`

and those are referenced here